File: src/main/java/com/leijendary/spring/microservicetemplate/model/LocalizedCopy.java
package com.leijendary.spring.microservicetemplate.model;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;

import static com.leijendary.spring.microservicetemplate.util.RequestContext.getLanguage;
import static java.util.Comparator.comparingInt;
import static java.util.stream.Collectors.toList;
import static javax.persistence.FetchType.EAGER;

@Data
@EqualsAndHashCode(callSuper = true)
@MappedSuperclass
public abstract class LocalizedCopy<T extends LocaleCopy> extends AbstractModel {

    @Id
    private long id;

    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    @ElementCollection(fetch = EAGER)
    @CollectionTable(joinColumns = @JoinColumn(name = "id"))
    private Set<T> translations = new HashSet<>();

    public T getTranslation() {
        final var language = getLanguage();
        final var sorted = translations
                .stream()
                .sorted(comparingInt(LocaleCopy::getOrdinal))
                .collect(toList());

        return sorted
                .stream()
                .filter(t -> t.getLanguage().equals(language))
                .findFirst()
                .orElse(sorted.iterator().next());
    }
}

File: src/main/kotlin/com/leijendary/spring/template/core/model/LocalizedCopy.kt
package com.leijendary.spring.template.core.model

import com.leijendary.spring.template.core.projection.UUIDProjection
import com.leijendary.spring.template.core.util.RequestContext.language
import java.io.Serializable

interface LocalizedCopy<T : LocaleCopy> : UUIDProjection, Serializable {
    val translations: Set<T>

    val translation: T
        get() {
            val language: String = language
            val sorted = translations.sortedBy { it.ordinal }

            return sorted.firstOrNull { it.language == language } ?: sorted.first()
        }
}
